File: java/src/test/java/com/azure/data/cosmos/serialization/hybridrow/unit/AssertThrowsException.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.data.cosmos.serialization.hybridrow.unit;

public final class AssertThrowsException {
    /**
     * Safely converts an object to a string, handling null values and null characters. Null
     * values are converted to "(null)". Null characters are converted to "\\0".
     *
     * @param input The object to convert to a string.
     * @return The converted string.
     */
    public static String ReplaceNulls(Object input) {
        String input1 = input == null ? null : input.toString();
        if (input1 == null) {
            return "(null)";
        }

        return Assert.ReplaceNullChars(input1);
    }

    /**
     * Tests whether the code specified by delegate <paramref name="action" /> throws exact given
     * exception of type <typeparamref name="T" /> (and not of derived type) and throws <code>
     * AssertFailedException
     * </code> if code does not throws exception or throws exception of type other than
     * <typeparamref name="T" />.
     *
     * @param action  Delegate to code to be tested and which is expected to throw exception.
     * @param message The message to include in the exception when <paramref name="action" /> does
     *                not throws exception of type <typeparamref name="T" />.
     *
     *                <typeparam name="T">Type of exception expected to be thrown.</typeparam>
     * @return The exception that was thrown.
     * @throws T:Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException Thrown if
     *                                                                              <paramref name="action" /> does
     *                                                                              not throws exception of type
     *                                                                              <typeparamref name="T" />.
     */
    public static <T extends RuntimeException> T ThrowsException(tangible.Action0Param action, String message) {
        return AssertThrowsException.ThrowsException(action, message, null);
    }

    /**
     * Tests whether the code specified by delegate <paramref name="action" /> throws exact given
     * exception of type <typeparamref name="T" /> (and not of derived type) and throws <code>
     * AssertFailedException
     * </code> if code does not throws exception or throws exception of type other than
     * <typeparamref name="T" />.
     *
     * @param action     Delegate to code to be tested and which is expected to throw exception.
     * @param message    The message to include in the exception when <paramref name="action" /> does
     *                   not throws exception of type <typeparamref name="T" />.
     * @param parameters An array of parameters to use when formatting <paramref name="message" />.
     *                   <typeparam name="T">Type of exception expected to be thrown.</typeparam>
     * @return The exception that was thrown.
     * @throws T:Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException Thrown if
     *                                                                              <paramref name="action" /> does
     *                                                                              not throws exception of type
     *                                                                              <typeparamref name="T" />.
     */
    public static <T extends RuntimeException> T ThrowsException(tangible.Action0Param action, String message,
                                                                 Object... parameters) {
        if (action == null) {
            throw new NullPointerException("action");
        }

        if (message == null) {
            throw new NullPointerException("message");
        }

        try {
            action.invoke();
        } catch (RuntimeException ex) {
            if (T.class != ex.getClass()) {
                Assert.Fail(String.format("Threw exception %3$s, but exception %2$s was expected. %1$s\nException " +
                    "Message: %4$s\nStack Trace: %5$s", AssertThrowsException.ReplaceNulls(message),
                    (Object)T.class.Name, ex.getClass().getSimpleName(), ex.getMessage(),
                    (Object)ex.StackTrace), parameters);
            }

            return (T)ex;
        }

        Assert.Fail(String.format("No exception thrown. %2$s exception was expected. %1$s",
            AssertThrowsException.ReplaceNulls(message), T.class.Name), parameters);

        return null;
    }

    /**
     * Tests whether the code specified by delegate <paramref name="action" /> throws exact given
     * exception of type <typeparamref name="T" /> (and not of derived type) and throws <code>
     * AssertFailedException
     * </code> if code does not throws exception or throws exception of type other than
     * <typeparamref name="T" />.
     *
     * @param action Delegate to code to be tested and which is expected to throw exception.
     *               <typeparam name="T">Type of exception expected to be thrown.</typeparam>
     * @return The exception that was thrown.
     * @throws T:Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException Thrown if
     *                                                                              <paramref name="action" /> does
     *                                                                              not throws exception of type
     *                                                                              <typeparamref name="T" />.
     */
    public static <T extends RuntimeException> T ThrowsException(tangible.Action0Param action) {
        return AssertThrowsException.ThrowsException(action, "", null);
    }
}

File: src/Serialization/HybridRow.Tests.Unit/AssertThrowsException.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

namespace Microsoft.Azure.Cosmos.Serialization.HybridRow.Tests.Unit
{
    using System;
    using Microsoft.VisualStudio.TestTools.UnitTesting;

    internal static class AssertThrowsException
    {
        /// <summary>
        /// Tests whether the code specified by delegate <paramref name="action" /> throws exact given
        /// exception of type <typeparamref name="T" /> (and not of derived type) and throws <code>
        /// AssertFailedException
        /// </code> if code does not throws exception or throws exception of type other than
        /// <typeparamref name="T" />.
        /// </summary>
        /// <param name="action">Delegate to code to be tested and which is expected to throw exception.</param>
        /// <typeparam name="T">Type of exception expected to be thrown.</typeparam>
        /// <exception cref="T:Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException">
        /// Thrown if
        /// <paramref name="action" /> does not throws exception of type <typeparamref name="T" />.
        /// </exception>
        /// <returns>The exception that was thrown.</returns>
        public static T ThrowsException<T>(Action action)
            where T : Exception
        {
            return AssertThrowsException.ThrowsException<T>(action, string.Empty, null);
        }

        /// <summary>
        /// Tests whether the code specified by delegate <paramref name="action" /> throws exact given
        /// exception of type <typeparamref name="T" /> (and not of derived type) and throws <code>
        /// AssertFailedException
        /// </code> if code does not throws exception or throws exception of type other than
        /// <typeparamref name="T" />.
        /// </summary>
        /// <param name="action">Delegate to code to be tested and which is expected to throw exception.</param>
        /// <param name="message">
        /// The message to include in the exception when <paramref name="action" /> does
        /// not throws exception of type <typeparamref name="T" />.
        /// </param>
        /// <typeparam name="T">Type of exception expected to be thrown.</typeparam>
        /// <exception cref="T:Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException">
        /// Thrown if
        /// <paramref name="action" /> does not throws exception of type <typeparamref name="T" />.
        /// </exception>
        /// <returns>The exception that was thrown.</returns>
        public static T ThrowsException<T>(Action action, string message)
            where T : Exception
        {
            return AssertThrowsException.ThrowsException<T>(action, message, null);
        }

        /// <summary>
        /// Tests whether the code specified by delegate <paramref name="action" /> throws exact given
        /// exception of type <typeparamref name="T" /> (and not of derived type) and throws <code>
        /// AssertFailedException
        /// </code> if code does not throws exception or throws exception of type other than
        /// <typeparamref name="T" />.
        /// </summary>
        /// <param name="action">Delegate to code to be tested and which is expected to throw exception.</param>
        /// <param name="message">
        /// The message to include in the exception when <paramref name="action" /> does
        /// not throws exception of type <typeparamref name="T" />.
        /// </param>
        /// <param name="parameters">An array of parameters to use when formatting <paramref name="message" />.</param>
        /// <typeparam name="T">Type of exception expected to be thrown.</typeparam>
        /// <exception cref="T:Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException">
        /// Thrown if
        /// <paramref name="action" /> does not throws exception of type <typeparamref name="T" />.
        /// </exception>
        /// <returns>The exception that was thrown.</returns>
        public static T ThrowsException<T>(Action action, string message, params object[] parameters)
            where T : Exception
        {
            if (action == null)
            {
                throw new ArgumentNullException(nameof(action));
            }

            if (message == null)
            {
                throw new ArgumentNullException(nameof(message));
            }

            try
            {
                action();
            }
            catch (Exception ex)
            {
                if (typeof(T) != ex.GetType())
                {
                    Assert.Fail(
                        string.Format(
                            "Threw exception {2}, but exception {1} was expected. {0}\nException Message: {3}\nStack Trace: {4}",
                            (object)AssertThrowsException.ReplaceNulls(message),
                            (object)typeof(T).Name,
                            (object)ex.GetType().Name,
                            (object)ex.Message,
                            (object)ex.StackTrace),
                        parameters);
                }

                return (T)ex;
            }

            Assert.Fail(
                string.Format(
                    "No exception thrown. {1} exception was expected. {0}",
                    AssertThrowsException.ReplaceNulls(message),
                    typeof(T).Name),
                parameters);

            return default;
        }

        /// <summary>
        /// Safely converts an object to a string, handling null values and null characters. Null
        /// values are converted to "(null)". Null characters are converted to "\\0".
        /// </summary>
        /// <param name="input">The object to convert to a string.</param>
        /// <returns>The converted string.</returns>
        internal static string ReplaceNulls(object input)
        {
            string input1 = input?.ToString();
            if (input1 == null)
            {
                return "(null)";
            }

            return Assert.ReplaceNullChars(input1);
        }
    }
}
